### PR TITLE
Integrate EIA 923 ER

### LINKED ISSUE #3677: Integrate EIA923 2023 Early Release Data
```[tasklist]
### Tasks
- [x] Archive data: https://zenodo.org/records/12656894
- [ ] Run through the steps in [Existing data Updates](https://catalystcoop-pudl.readthedocs.io/en/nightly/dev/existing_data_updates.html)
```

### PR DESCRIPTION
**Pull request by aesharpe**
# Overview

Closes #3677.

## What problem does this address?

Adds the new EIA 923 ER data.

## What did you change?

**1. Added the DOI for the new data.**

**2. Updated all the `package_data` files to reflect the ER data files.**

**3. Fixed a quirk in the `core_eia923__monthly_generation_fuel_nuclear` table that was causing the `out_eia923__generation_fuel_combined` table to fail**

```
ValueError: Found inhomogeneous non-data cols while aggregating nuclear generation. Non-data cols: ['fuel_type_code_aer', 'fuel_type_code_pudl', 'data_maturity']
```
The failure was caused by`None` values in the `fuel_type_code_aer` column in the `core_eia923__monthly_generation_fuel_nuclear` table. This was tripping a `ValueError` in the creation of the `out_eia923__generation_fuel_combined` table: 

```
gfn_gb = core_eia923__monthly_generation_fuel_nuclear.groupby(primary_key)
# Ensure that all non-data columns are homogeneous within groups
if gfn_gb[non_data_cols].nunique().ne(1).any(axis=None):
    raise ValueError(
        "Found inhomogeneous non-data cols while aggregating nuclear generation. "
        f"Non-data cols: {non_data_cols}"
    )
```
I added a line to the transform function for `core_eia923__monthly_generation_fuel_nuclear` that converted any None values to NUC (so long as the `energy_source_code` was NUC) and that fixed the problem.

**4. Update min-max rows validation test.**
There are some newly harvested plant_id-year combos that add rows to the `plants_eia860` and `boil_eia860` tables.

**5. Update release notes.**

# Testing

How did you make sure this worked? How can a reviewer verify this?

```[tasklist]
# To-do list
- [x] If updating analyses or data processing functions: make sure to update or write data validation tests (e.g.,  `test_minmax_rows()`)
- [x] Update the [release notes](../docs/release_notes.rst): reference the PR and related issues.
- [ ] Ensure docs build, unit & integration tests, and test coverage pass locally with `make pytest-coverage` (otherwise the merge queue may reject your PR)
- [x] Review the PR yourself and call out any questions or issues you have
- [x] For minor ETL changes or data additions, once `make pytest-coverage` passes, make sure you have a fresh full PUDL DB downloaded locally, materialize new/changed assets and all their downstream assets and [run relevant data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation) using `pytest` and `--live-dbs`.
- [ ] For significant ETL, data coverage or analysis changes, once `make pytest-coverage` passes, ensure the full ETL runs locally and [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation) using `make pytest-validate` (a ~10 hour run). If you can't run this locally, run the `build-deploy-pudl` GitHub Action (or ask someone with permissions to). Then, check the logs on the `#pudl-deployments` Slack channel or `gs://builds.catalyst.coop`.
```

### CODE REVIEW

**Comment by aesharpe** [comment]
Getting a foreign key failure and I'm bad at debugging them so soliciting advice from @e-belfer 

Ran `pytest test/integration --live-dbs`

```
FAILED test/integration/etl_test.py::test_pudl_engine - pudl.etl.check_foreign_keys.ForeignKeyErrors: Foreign key error for table: _out_eia__yearly_generators -- core_eia860__scd_generators (plant_id_eia, gener...
FAILED test/integration/record_linkage_test.py::test_classify_plants_ferc1 - mlflow.exceptions.MlflowException: Detected out-of-date database schema (found version 867495a8f9d4, but expected 5b0e9adcef9c). Take a backup of your dat...
```

```
test/integration/etl_test.py:36: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

engine = Engine(sqlite:////Users/austensharpe/Desktop/pudl_workspace/pudl_output/pudl.sqlite)

    def check_foreign_keys(engine: sa.Engine):
        """Check foreign key relationships in the database.
    
        The order assets are loaded into the database will not satisfy foreign key
        constraints so we can't enable foreign key constraints. However, we can
        check for foreign key failures once all of the data has been loaded into
        the database using the `foreign_key_check` and `foreign_key_list` PRAGMAs.
    
        You can learn more about the PRAGMAs in the `SQLite docs
        <https://www.sqlite.org/pragma.html#pragma_foreign_key_check>`__.
    
        Raises:
            ForeignKeyErrors: if data in the database violate foreign key constraints.
        """
        logger.info(f"Running foreign key check on {engine.url} database.")
        with engine.begin() as con:
            fk_errors = pd.read_sql_query("PRAGMA foreign_key_check;", con)
    
        if not fk_errors.empty:
            # Merge in the actual FK descriptions
            tables_with_fk_errors = fk_errors.table.unique().tolist()
            table_foreign_keys = pd.concat(
                [_get_fk_list(engine, table) for table in tables_with_fk_errors]
            )
    
            fk_errors_with_keys = fk_errors.merge(
                table_foreign_keys,
                how="left",
                on=["parent", "fkid", "table"],
                validate="m:1",
            )
    
            errors = []
            # For each foreign key error, raise a ForeignKeyError
            for (
                table_name,
                parent_name,
                parent_fk,
            ), parent_fk_df in fk_errors_with_keys.groupby(["table", "parent", "fk"]):
                errors.append(
                    ForeignKeyError(
                        child_table=table_name,
                        parent_table=parent_name,
                        foreign_key=parent_fk,
                        rowids=parent_fk_df["rowid"].values,
                    )
                )
>           raise ForeignKeyErrors(errors)
E           pudl.etl.check_foreign_keys.ForeignKeyErrors: Foreign key error for table: _out_eia__yearly_generators -- core_eia860__scd_generators (plant_id_eia, generator_id, report_date) -- on rows [551972 566041 566042]
E           
E           Foreign key error for table: out_eia923__boiler_fuel -- core_eia__entity_boilers (plant_id_eia, boiler_id) -- on rows [1623953 1623954 1623955 1623956 1623957 1623958 1623959 1623960 1623961
E            1623962 1623963 1623964 1674125 1674126 1674127 1674128 1674129 1674130
E            1674131 1674132 1674133 1674134 1674135 1674136]
E           
E           Foreign key error for table: out_eia923__generation -- core_eia__entity_generators (plant_id_eia, generator_id) -- on rows [699155 699156 699157 699158 699159 699160 699161 699162 699163 699164
E            699165 699166 699167 699168 699169 699170 699171 699172 699173 699174
E            699175 699176 699177 699178]
E           
E           Foreign key error for table: out_eia923__monthly_boiler_fuel -- core_eia__entity_boilers (plant_id_eia, boiler_id) -- on rows [   3277    3278    3279    3280    3281    3282    3283    3284    3285
E               3286    3287    3288 1284764 1284765 1284766 1284767 1284768 1284769
E            1284770 1284771 1284772 1284773 1284774 1284775]
E           
E           Foreign key error for table: out_eia923__monthly_generation -- core_eia__entity_generators (plant_id_eia, generator_id) -- on rows [278918 278919 278920 278921 278922 278923 278924 278925 278926 278927
E            278928 278929 278930 278931 278932 278933 278934 278935 278936 278937
E            278938 278939 278940 278941]
E           
E           Foreign key error for table: out_eia923__yearly_boiler_fuel -- core_eia860__scd_boilers (plant_id_eia, boiler_id, report_date) -- on rows [   255   4514   8238   8248   8724   8734  17623  17633  19272  19282
E             23724  23735  33302  33312  33322  34710  38477  38488  39061  39071
E             39081  39091  39101  39111  49496  51398  51409  52200  52211  52446
E             52720  52932  54028  54038  54048  54058  54172  54182  54791  54801
E             58259  58305  58316  60418  60428  60438  60448  60487  68513  68523
E             68560  68571  68598  68609  70423  70433  70443  70453  71269  71279
E             71289  71299  72062  72073  72916  72927  72967  76523  76533  76543
E             76553  76760  76770  76876  76886  76896  77081  79161  79166  80854
E             80864  80872  80882  80892  81843  82239  82246  82253  82260  84023
E             84034  84072  84083  84223  84234  84245  86620  86858  86869  86880
E             86891  86994  87004  87009  87019  91352  91377  93163  93167  96705
E             96716  96727  96738  96817  96827  96836  96846  96992  97003  99500
E             99511 101054 101065 101240 101251 102704 102714 104477 104488 104499
E            104510 104575 104586 106447 106457 106466 106476 107102 107113 107425
E            107436 108042 108600 108611 108622 109131 109141 109214 109225 109295
E            109306 109313 109355 109917 109927 109937 109947 113030 113040 113216
E            113226 113236 113246 113899 113909 114109 114375 114624 114634 117096
E            117156 117165 118013 118023 118033 118043 118053 118063 118073 118165
E            126940 127825 127835 127845 128285 128295 128645 128656 129299 130221
E            130586 130596 131197 131207 131217 131224 132148 132156 133640 133826
E            134110 134121 134607 134615 134906 134925 134932 135858 135863 135868
E            135897 135904 135908 135912 135917 135921 135925 135934 135938 135942]
E           
E           Foreign key error for table: out_eia923__yearly_generation -- core_eia860__scd_generators (plant_id_eia, generator_id, report_date) -- on rows [57446]
E           
E           Foreign key error for table: out_eia__yearly_boilers -- core_eia860__scd_boilers (plant_id_eia, boiler_id, report_date) -- on rows [80129 80327 80513 80535 80922 80990 81153 81553 81554 81555 81604 81818
E            81860 81861 81862 82277 82362 82401 82417 82435 82446 82505 82506 82513
E            82514 82537 82538 82646 82649 82705 82706 82709 82891 82893 82895 82933
E            82934 82935 82936 82965 82966 83000 83044 83045 83163 83164 83174 83175
E            83181 83182 83183 83191 83240 83280 83281 83282 83311 83325 83326 83377
E            83379 83384 83451 83458 83459 83465 83466 83578 83580 83624 83625 83729
E            83730 83734 83735 83740 83741 83836 83837 83883 83886 83915 83940 83941
E            83944 84013 84014 84042 84043 84062 84075 84106 84107 84108 84126 84127
E            84131 84132 84134 84147 84148 84232 84240 84241 84271 84284 84297 84314
E            84315 84500 84505 84506 84566 84567 84568 84569 84570 84571 84572 84580
E            85144 85195 85196 85218 85234 85235 85273 85321 85339 85340 85375 85376
E            85377 85378 85433 85434 85527 85536 85551 85552 85588 85589 85615 85616
E            85617 85752 85753 85754 85765 85766 85767 85768 85769 85770 85771 85772
E            85805 86568 87628]

src/pudl/etl/check_foreign_keys.py:208: ForeignKeyErrors
```

File: src/pudl/metadata/codes.py [hunk 1]
```diff
@@ -1681,7 +1681,10 @@
                 ("WT", "wind_onshore", "Wind Turbine, Onshore"),
             ],
         ).convert_dtypes(),
-        "code_fixes": {"ic": "IC"},  # there is literally one 'ic' from 2002.
+        "code_fixes": {
+            "ic": "IC",
+            "WY": "WT",
```
Comment (e-belfer) [inline src/pudl/metadata/codes.py:1686]: How do you know this?

**Review by e-belfer — CHANGES_REQUESTED** [review]
In looking at the raw extraction logs for 2023, it looks like there are a bunch of columns in tables that were included in the monthly data that may have been renamed or added in the ER data and that we may be missing.

Was able to materialize `raw_eia923` and `core_eia923` assets with no errors, but these warnings mean we are effectively missing data for this year for these columns!

```
2024-07-17 14:23:08 [ WARNING] catalystcoop.pudl.extract.extractor:190 boiler_fuel/2023: Extra columns found in extracted table:
{'balancing_authority_code', 'respondent_frequency', 'early_release_data_july_2024_not_fully_edited_use_with_caution_do_not_aggregate_to_state_regional_or_national_totals'}
2024-07-17 14:23:08 [ WARNING] catalystcoop.pudl.extract.extractor:195 boiler_fuel/2023: Expected columns not found in extracted table:
{'balancing_authority_code_eia'}

2024-07-17 14:23:10 [ WARNING] catalystcoop.pudl.extract.extractor:190 energy_storage/2023: Extra columns found in extracted table:
{'respondent_frequency', 'mer_fuel_type_code', 'early_release_data_july_2024_not_fully_edited_use_with_caution_do_not_aggregate_to_state_regional_or_national_totals'}
2024-07-17 14:23:10 [ WARNING] catalystcoop.pudl.extract.extractor:195 energy_storage/2023: Expected columns not found in extracted table:
{'fuel_type_code_aer'}

2024-07-17 14:23:15 [ WARNING] catalystcoop.pudl.extract.extractor:190 fuel_receipts_costs/2023: Extra columns found in extracted table:
{'balancing_authority_code', 'reporting_frequency', 'early_release_data_july_2024_not_fully_edited_use_with_caution_do_not_aggregate_to_state_regional_or_national_totals'}
2024-07-17 14:23:15 [ WARNING] catalystcoop.pudl.extract.extractor:195 fuel_receipts_costs/2023: Expected columns not found in extracted table:
{'balancing_authority_code_eia'}

2024-07-17 14:23:20 [ WARNING] catalystcoop.pudl.extract.extractor:190 generation_fuel/2023: Extra columns found in extracted table:
{'respondent_frequency', 'mer_fuel_type_code', 'early_release_data_july_2024_not_fully_edited_use_with_caution_do_not_aggregate_to_state_regional_or_national_totals'}
2024-07-17 14:23:20 [ WARNING] catalystcoop.pudl.extract.extractor:195 generation_fuel/2023: Expected columns not found in extracted table:
{'fuel_type_code_aer'}

2024-07-17 14:23:20 [ WARNING] catalystcoop.pudl.extract.extractor:190 generator/2023: Extra columns found in extracted table:
{'balancing_authority_code', 'respondent_frequency', 'early_release_data_july_2024_not_fully_edited_use_with_caution_do_not_aggregate_to_state_regional_or_national_totals'}
2024-07-17 14:23:20 [ WARNING] catalystcoop.pudl.extract.extractor:195 generator/2023: Expected columns not found in extracted table:
{'balancing_authority_code_eia'}
```

File: src/pudl/metadata/codes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aesharpe) [reply src/pudl/metadata/codes.py:1686]: I believe I looked at other years of the same plant

**Comment by aesharpe** [comment]
The `mer_fuel_type_code` column from the `energy_storage` table is kind of annoying. It was previously `aer_fuel_type_code` which stands for Annual Energy Review as opposed to Monthly Energy Review. Im not exactly sure what this means in the context of this table. Maybe it's the fuel type taken from the most recent monthly review? as opposed to annual? I'm tempted to lump it in with the `aer_fuel_type_code` because it feels silly for them to be separate columns. But, at the same time, it is sort of different?

File: src/pudl/metadata/codes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aesharpe) [reply src/pudl/metadata/codes.py:1686]: There was only one instance of this.

**Comment by e-belfer** [comment]
> The mer_fuel_type_code column from the energy_storage and generation_fuel tables is kind of annoying. It was previously aer_fuel_type_code which stands for Annual Energy Review as opposed to Monthly Energy Review. Im not exactly sure what this means in the context of this table. Maybe it's the fuel type taken from the most recent monthly review as opposed to annual review? I'm tempted to lump it in with the aer_fuel_type_codecolumn because it feels silly for them to be separate. But, at the same time, it is sort of different? What do you think?

I was wondering if this is just an ER thing, but it's not. It's because AER has been suspended: https://www.eia.gov/totalenergy/data/annual/

The AER fuel type code from 2022 ER data is defined as this in the EIA metadata:

> A partial aggregation of the reported fuel type codes into larger categories used by EIA in, for example, the Annual Energy Review (AER).Two or three letter alphanumeric.  See the Fuel Code table (Table 5), below:

The 2023 MER fuel type code is defined as this in the EIA metadata:

> A partial aggregation of the reported fuel type codes into larger categories used by EIA in, for example, the Monthly Energy Review (MER).Two or three letter alphanumeric.  See the Fuel Code table (Table 5), below:

Do you feel confident that the categories themselves haven't changed, and do you know if the fuel type codes in MER and AER are the same? If so, I think we could combine these and just say "in the Monthly Energy Reviews or the former Annual Energy Reviews" in the field description, and possibly consider renaming the column towards mer?

**Comment by aesharpe** [comment]
> Do you feel confident that the categories themselves haven't changed, and do you know if the fuel type codes in MER and AER are the same? If so, I think we could combine these and just say "in the Monthly Energy Reviews or the former Annual Energy Reviews" in the field description, and possibly consider renaming the column towards mer?

The fuel type codes are the same, and I'm almost positive the categories haven't changed. I'll change the description.

**Comment by aesharpe** [comment]
I'm getting a really weird alembic error when I try and upgrade the column names to have the suffix `_mer` instead of `_aer`. All is fine until I update the `core_eia__codes_fuel_types_aer` code table. For some reason when I change that to `_mer` (there are three instances of that -- one in `CODE_METADATA` in `codes/metadata.py` and two in `RESOURCE_METADATA` in `metadata/resources/eia.py`.

After I run `alembic revision --autogenerate` I get the following code: 

```
"""AER to MER x2

Revision ID: 9625b1bd89e3
Revises: 08901b7bc357
Create Date: 2024-07-19 12:12:51.592927

"""
from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision = '9625b1bd89e3'
down_revision = '08901b7bc357'
branch_labels = None
depends_on = None


def upgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('core_eia__codes_fuel_types_mer',
    sa.Column('code', sa.Text(), nullable=False, comment='Originally reported short code.'),
    sa.Column('description', sa.Text(), nullable=True, comment='Long human-readable description of the meaning of a code/label.'),
    sa.PrimaryKeyConstraint('code', name=op.f('pk_core_eia__codes_fuel_types_mer'))
    )
    op.drop_table('core_eia__codes_fuel_types_aer')
    with op.batch_alter_table('core_eia923__monthly_generation_fuel', schema=None) as batch_op:
        batch_op.drop_constraint('fk_core_eia923__monthly_generation_fuel_fuel_type_code_mer_core_eia__codes_fuel_types_aer', type_='foreignkey')
        batch_op.create_foreign_key(batch_op.f('fk_core_eia923__monthly_generation_fuel_fuel_type_code_mer_core_eia__codes_fuel_types_mer'), 'core_eia__codes_fuel_types_mer', ['fuel_type_code_mer'], ['code'])

    with op.batch_alter_table('core_eia923__monthly_generation_fuel_nuclear', schema=None) as batch_op:
        batch_op.drop_constraint('fk_core_eia923__monthly_generation_fuel_nuclear_fuel_type_code_mer_core_eia__codes_fuel_types_aer', type_='foreignkey')
        batch_op.create_foreign_key(batch_op.f('fk_core_eia923__monthly_generation_fuel_nuclear_fuel_type_code_mer_core_eia__codes_fuel_types_mer'), 'core_eia__codes_fuel_types_mer', ['fuel_type_code_mer'], ['code'])

    with op.batch_alter_table('out_eia923__generation_fuel_combined', schema=None) as batch_op:
        batch_op.drop_constraint('fk_out_eia923__generation_fuel_combined_fuel_type_code_mer_core_eia__codes_fuel_types_aer', type_='foreignkey')
        batch_op.create_foreign_key(batch_op.f('fk_out_eia923__generation_fuel_combined_fuel_type_code_mer_core_eia__codes_fuel_types_mer'), 'core_eia__codes_fuel_types_mer', ['fuel_type_code_mer'], ['code'])

    # ### end Alembic commands ###


def downgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    with op.batch_alter_table('out_eia923__generation_fuel_combined', schema=None) as batch_op:
        batch_op.drop_constraint(batch_op.f('fk_out_eia923__generation_fuel_combined_fuel_type_code_mer_core_eia__codes_fuel_types_mer'), type_='foreignkey')
        batch_op.create_foreign_key('fk_out_eia923__generation_fuel_combined_fuel_type_code_mer_core_eia__codes_fuel_types_aer', 'core_eia__codes_fuel_types_aer', ['fuel_type_code_mer'], ['code'])

    with op.batch_alter_table('core_eia923__monthly_generation_fuel_nuclear', schema=None) as batch_op:
        batch_op.drop_constraint(batch_op.f('fk_core_eia923__monthly_generation_fuel_nuclear_fuel_type_code_mer_core_eia__codes_fuel_types_mer'), type_='foreignkey')
        batch_op.create_foreign_key('fk_core_eia923__monthly_generation_fuel_nuclear_fuel_type_code_mer_core_eia__codes_fuel_types_aer', 'core_eia__codes_fuel_types_aer', ['fuel_type_code_mer'], ['code'])

    with op.batch_alter_table('core_eia923__monthly_generation_fuel', schema=None) as batch_op:
        batch_op.drop_constraint(batch_op.f('fk_core_eia923__monthly_generation_fuel_fuel_type_code_mer_core_eia__codes_fuel_types_mer'), type_='foreignkey')
        batch_op.create_foreign_key('fk_core_eia923__monthly_generation_fuel_fuel_type_code_mer_core_eia__codes_fuel_types_aer', 'core_eia__codes_fuel_types_aer', ['fuel_type_code_mer'], ['code'])

    op.create_table('core_eia__codes_fuel_types_aer',
    sa.Column('code', sa.TEXT(), nullable=False),
    sa.Column('description', sa.TEXT(), nullable=True),
    sa.PrimaryKeyConstraint('code', name='pk_core_eia__codes_fuel_types_aer')
    )
    op.drop_table('core_eia__codes_fuel_types_mer')
    # ### end Alembic commands ###

```

However, when I run `alembic upgrade head` I get the following error: 

```
INFO  [root] alembic config.sqlalchemy.url: sqlite:////Users/austensharpe/Desktop/pudl_workspace/pudl_output/pudl.sqlite
Traceback (most recent call last):
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/bin/alembic", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/config.py", line 636, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/config.py", line 626, in main
    self.run_cmd(cfg, options)
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/config.py", line 603, in run_cmd
    fn(
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/command.py", line 406, in upgrade
    script.run_env()
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/script/base.py", line 582, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/util/pyfiles.py", line 95, in load_python_file
    module = load_module_py(module_id, path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/util/pyfiles.py", line 113, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/Users/austensharpe/Desktop/Repos/pudl/migrations/env.py", line 81, in <module>
    run_migrations_online()
  File "/Users/austensharpe/Desktop/Repos/pudl/migrations/env.py", line 75, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/runtime/environment.py", line 946, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/runtime/migration.py", line 628, in run_migrations
    step.migration_fn(**kw)
  File "/Users/austensharpe/Desktop/Repos/pudl/migrations/versions/9625b1bd89e3_aer_to_mer_x2.py", line 27, in upgrade
    with op.batch_alter_table('core_eia923__monthly_generation_fuel', schema=None) as batch_op:
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/operations/base.py", line 398, in batch_alter_table
    impl.flush()
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/alembic/operations/batch.py", line 142, in flush
    existing_table = Table(
                     ^^^^^^
  File "<string>", line 2, in __new__
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/util/deprecations.py", line 281, in warned
    return fn(*args, **kwargs)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/sql/schema.py", line 432, in __new__
    return cls._new(*args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/sql/schema.py", line 486, in _new
    with util.safe_reraise():
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/sql/schema.py", line 482, in _new
    table.__init__(name, metadata, *args, _no_init=False, **kw)
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/sql/schema.py", line 862, in __init__
    self._autoload(
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/sql/schema.py", line 894, in _autoload
    conn_insp.reflect_table(
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/engine/reflection.py", line 1570, in reflect_table
    self._reflect_fk(
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/engine/reflection.py", line 1768, in _reflect_fk
    sa_schema.Table(
  File "<string>", line 2, in __new__
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/util/deprecations.py", line 281, in warned
    return fn(*args, **kwargs)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/sql/schema.py", line 432, in __new__
    return cls._new(*args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/sql/schema.py", line 486, in _new
    with util.safe_reraise():
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/sql/schema.py", line 482, in _new
    table.__init__(name, metadata, *args, _no_init=False, **kw)
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/sql/schema.py", line 862, in __init__
    self._autoload(
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/sql/schema.py", line 894, in _autoload
    conn_insp.reflect_table(
  File "/Users/austensharpe/mambaforge/envs/pudl-dev/lib/python3.12/site-packages/sqlalchemy/engine/reflection.py", line 1538, in reflect_table
    raise exc.NoSuchTableError(table_name)
sqlalchemy.exc.NoSuchTableError: core_eia__codes_fuel_types_aer
```

**Comment by zaneselvans** [comment]
Does your current PUDL DB contain a table named `core_eia__codes_fuel_types_aer`?

**Comment by aesharpe** [comment]
> Does your current PUDL DB contain a table named `core_eia__codes_fuel_types_aer`?

It does, but when I run `alembic upgrade head` it switches the name to `core_eia__codes_fuel_types_mer` and then fails and shows me the above error. So the upgrade doesn't work, but the table name gets changed.

**Comment by zaneselvans** [comment]
What have you already tried to fix it?

**Comment by zaneselvans** [comment]
I was able to resolve the alembic issue. Searching through the code base, I'm not seeing any reference to `fuel_types_mer` and I suspect that's probably the way to go -- it's been `fuel_type_aer` for many years, and if all the codes are the same and things are in flux with the AER/MER/AEO etc. then retroactively changing all the past years of data to say "MER" seems a little weird or premature.

File: src/pudl/transform/eia923.py [hunk 2]
```diff
@@ -592,6 +592,11 @@ def gen_fuel_nuclear(gen_fuel_nuke: pd.DataFrame) -> pd.DataFrame:
         "ST"
     )
 
+    # All nuclear plants use nuclear fuel
+    gen_fuel_nuke.loc[
```
[the remark this answers is not in the document]
Comment (zaneselvans) [reply src/pudl/transform/eia923.py:596]: Weirdly I have seen a handful of nuclear plants report using small amounts of liquid fuels. Maybe to power backup generators?

**Comment by zaneselvans** [comment]
I did a full ETL locally and ran the validation tests overnight and got some row count discrepancies:

```
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-bga_eia860-151860-151860-151860] - ValueError: bga_eia860: found 153487 rows, expected 151860. Off by 1.071%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-bga_eia860-151860-151860-151860] - ValueError: bga_eia860: found 153487 rows, expected 151860. Off by 1.071%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-gen_eia923-None-5494698-459691] - ValueError: gen_eia923: found 459711 rows, expected 459691. Off by 0.004%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-hr_by_unit-406067-33946] - ValueError: hr_by_unit: found 34258 rows, expected 33946. Off by 0.919%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-hr_by_gen-634378-52978] - ValueError: hr_by_gen: found 53530 rows, expected 52978. Off by 1.042%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-fuel_cost-634378-52978] - ValueError: fuel_cost: found 53530 rows, expected 52978. Off by 1.042%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-bga_eia860-151860-151860-151860] - ValueError: bga_eia860: found 153487 rows, expected 151860. Off by 1.071%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-capacity_factor-5494698-459691] - ValueError: capacity_factor: found 459711 rows, expected 459691. Off by 0.004%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-mcoe-5495136-459726] - ValueError: mcoe: found 459746 rows, expected 459726. Off by 0.004%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-gen_eia923-None-5494698-459691] - ValueError: gen_eia923: found 5494932 rows, expected 5494698. Off by 0.004%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-mcoe-5495136-459726] - ValueError: mcoe: found 5495370 rows, expected 5495136. Off by 0.004%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-hr_by_unit-406067-33946] - ValueError: hr_by_unit: found 409787 rows, expected 406067. Off by 0.916%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-hr_by_gen-634378-52978] - ValueError: hr_by_gen: found 640960 rows, expected 634378. Off by 1.038%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-fuel_cost-634378-52978] - ValueError: fuel_cost: found 640960 rows, expected 634378. Off by 1.038%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-capacity_factor-5494698-459691] - ValueError: capacity_factor: found 5494932 rows, expected 5494698. Off by 0.004%, allowed margin of 0.000%
```

(also the FK constraints failed, but that's a FERC-714 problem that came from another branch)

**Comment by e-belfer** [comment]
@zaneselvans I don't think things are a bit in flux, it's been superseded entirely by MER.

**Comment by aesharpe** [comment]
I took a closer look at the `energy_source_code` (raw col: `reported_fuel_type_code`) and `fuel_type_code_aer / _mer` columns. `fuel_type_code_aer` is an aggregation of `energy_source_code` with one exception. `energy_source_code` will report `WAT` for hydro and `fuel_type_code_aer` will report either `HYC` or `HPS` for conventional vs. pumped storage. For this reason I kept both columns in the final table. 

I changed the `_aer` suffix to `_agg` instead of `_mer`. I also wiped the alembic migrations because I kept running into the same error as above. Confirmed this decision with @e-belfer.

**Review by e-belfer — APPROVED** [review]
Was able to materialize `raw_eia923` and `core_eia923` assets as expected. If the validation tests all pass on your end, should be good to go as far as I'm concerned!

**Pull request merged**